src/wallet/mod.rs

```diff
@@ -366,67 +332,236 @@ where
+                    .derived_descriptor(&self.secp)
+                    .expect("the descritpor cannot need hardened derivation")
+                    .script_pubkey();
+
+                self.spk_tracker.add_spk((keychain, index), spk);
             }
         }
-        Ok(new_addresses_cached)
+        true
+    }
+
+    /// Derive all addresses up to and including `index`.
+    pub fn ensure_derived_up_to(&mut self, keychain: KeychainKind, index: u32) -> bool {
+        let descriptor = self.get_descriptor_for_keychain(keychain).clone();
+        let end = match descriptor.has_wildcard() {
+            false => 0,
+            true => index,
+        };
+
+        let next_to_derive = self.next_derivation_index(keychain);
+        if next_to_derive > end {
+            return false;
+        }
+
+        for index in next_to_derive..=end {
+            let spk = descriptor
+                .at_derivation_index(index)
+                .derived_descriptor(&self.secp)
+                .expect("the descritpor cannot need hardened derivation")
+                .script_pubkey();
+
+            self.spk_tracker.add_spk((keychain, index), spk);
+        }
+
+        true
     }
 
     /// Return whether or not a `script` is part of this wallet (either internal or external)
-    pub fn is_mine(&self, script: &Script) -> Result<bool, Error> {
-        self.database.borrow().is_mine(script)
+    pub fn is_mine(&self, script: &Script) -> bool {
+        self.spk_tracker.index_of_spk(script).is_some()
     }
 
     /// Return the list of unspent outputs of this wallet
     ///
     /// Note that this method only operates on the internal database, which first needs to be
     /// [`Wallet::sync`] manually.
-    pub fn list_unspent(&self) -> Result<Vec<LocalUtxo>, Error> {
-        Ok(self
-            .database
-            .borrow()
-            .iter_utxos()?
-            .into_iter()
-            .filter(|l| !l.is_spent)
-            .collect())
+    pub fn list_unspent(&self) -> Vec<LocalUtxo> {
+        self.spk_tracker
+            .iter_unspent(&self.sparse_chain, &self.tx_graph)
+            .map(|((keychain, _), txo)| LocalUtxo {
+                outpoint: txo.outpoint,
+                txout: txo.txout,
+                keychain,
+                is_spent: false,
+            })
+            .collect()
+    }
+
+    /// Iterate over all checkpoints.
+    pub fn iter_checkpoints(&self) -> &BTreeMap<u32, BlockHash> {
+        self.sparse_chain.checkpoints()
+    }
+
+    /// Range checkpoints
+    pub fn range_checkpoints<'a>(
+        &'a self,
+        range: impl RangeBounds<u32> + 'a,
+    ) -> impl DoubleEndedIterator<Item = BlockId> + 'a {
+        self.sparse_chain.range_checkpoints(range)
+    }
+
+    /// Returns the latest checkpoint.
+    pub fn latest_checkpoint(&self) -> Option<BlockId> {
+        self.sparse_chain.latest_checkpoint()
+    }
+
+    /// Inserts a transaction into the graph if the transaction is deemed relevant.
```

Comment:
What does it mean for a tx to be relevant? To me, it seems that it means:
- it's a wallet transaction
- we have never seen it before

Is this correct?

Reply:
This method was included by accident. Sorry! It was something sketched out earlier that won't ever exist I think.